crypto_adapters/t_cose_openssl_crypto.c

```diff
@@ -455,42 +586,54 @@ t_cose_crypto_verify(const int32_t                cose_algorithm_id,
     unsigned               key_size;
-    MakeUsefulBufOnStack(  der_format_buffer, T_COSE_MAX_SIG_SIZE + DER_SIG_ENCODE_OVER_HEAD);
+
+    /* This buffer is used to convert COSE ECDSA signature to DER format,
+     * before it can be consumed by OpenSSL. When RSA signatures are
+     * selected the buffer is unused.
+     */
+    MakeUsefulBufOnStack(  der_format_buffer, T_COSE_MAX_ECDSA_SIG_SIZE + DER_SIG_ENCODE_OVER_HEAD);
     struct q_useful_buf_c  der_format_signature;
 
     /* This implementation doesn't use any key store with the ability
      * to look up a key based on kid. */
     (void)kid;
 
-    if(!t_cose_algorithm_is_ecdsa(cose_algorithm_id)) {
+    if(!t_cose_algorithm_is_ecdsa(cose_algorithm_id) &&
+       !t_cose_algorithm_is_rsassa_pss(cose_algorithm_id)) {
         return_value = T_COSE_ERR_UNSUPPORTED_SIGNING_ALG;
         goto Done;
     }
 
     /* Get the verification key in an EVP_PKEY structure which is what
-     * is needed for sig verification. This also gets the key size
-     * which is needed to convert the format of the signature. */
-    return_value = key_convert_and_size(verification_key,
-                                        &verification_key_evp,
-                                        &key_size);
+     * is needed for sig verification. */
+    return_value = key_convert(verification_key, &verification_key_evp);
     if(return_value != T_COSE_SUCCESS) {
         goto Done;
     }
 
-    /* Unfortunately the officially supported OpenSSL API supports
-     * only DER-encoded signatures so the COSE format signatures must
-     * be converted to DER for verification. This requires a temporary
-     * buffer and a fair bit of work inside signature_cose_to_der().
-     */
-    return_value = signature_cose_to_der(key_size,
-                                         cose_signature,
-                                         der_format_buffer,
-                                        &der_format_signature);
-    if(return_value) {
+    if (t_cose_algorithm_is_ecdsa(cose_algorithm_id)) {
+        /* Unfortunately the officially supported OpenSSL API supports
+         * only DER-encoded signatures so the COSE format ECDSA signatures must
+         * be converted to DER for verification. This requires a temporary
+         * buffer and a fair bit of work inside ecdsa_signature_cose_to_der().
+         */
+        return_value = ecdsa_signature_cose_to_der(verification_key_evp,
+                                                   cose_signature,
+                                                   der_format_buffer,
+                                                   &der_format_signature);
+        if(return_value) {
+          goto Done;
+        }
+    } else if (t_cose_algorithm_is_rsassa_pss(cose_algorithm_id)) {
+        /* COSE RSA signatures are already in the format OpenSSL
+         * expects, they can be used without any re-encoding.
+         */
+        der_format_signature = cose_signature;
```

Comment:
Strictly speaking, the RSA signature is not DER, so the variable name is slightly misleading.

Reply:
Good point. Renamed it to `openssl_signature` and added a comment near the declaration to exclain.